lib/SILOptimizer/PassManager/PassManager.cpp

```diff
@@ -523,6 +523,48 @@ runFunctionPasses(unsigned FromTransIdx, unsigned ToTransIdx) {
       ++Entry.PipelineIdx;
     }
     clearRestartPipeline();
+
+    if (TailIdx == (FunctionWorklist.size() - 1))  {
+      // No new functions to process
+      continue;
+    }
+
+    // Collect new functions added to the worklist
+    std::vector<SILFunction *> NewFunctions;
+    NewFunctions.reserve(FunctionWorklist.size() - (TailIdx + 1));
+    std::transform(FunctionWorklist.begin() + (TailIdx + 1),
+                   FunctionWorklist.end(), std::back_inserter(NewFunctions),
+                   [](const WorklistEntry &w) { return w.F; });
+
+    // Compute the bottom up order of the new functions and the callees in it
+    BottomUpFunctionOrder SubBottomUpOrder(std::move(NewFunctions), BCA);
+    auto NewFunctionsBottomUp = SubBottomUpOrder.getFunctions();
+
+    // Remove all the functions in the new bottom up order from FunctionWorklist
+    llvm::DenseMap<SILFunction*, WorklistEntry> FunctionsToReorder;
+    auto RemoveFn = [&FunctionsToReorder,
+                     &NewFunctionsBottomUp](WorklistEntry Entry) {
+      if (std::find(NewFunctionsBottomUp.begin(), NewFunctionsBottomUp.end(),
```

Comment:
To avoid the linear search you could first add all new functions to a SmallPtrSet and then do a fast lookup here.